src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -1316,20 +1318,110 @@ private static string IndentSource(string source, int numIndentations)
-        {
-            switch (value)
+            private string GetParamDefaultValueAsString(object? value, Type type, string typeRef)
+            {
+                if (value == null)
+                {
+                    return $"default({typeRef})";
+                }
+
+                if (type.IsEnum)
+                {
+                    // Roslyn gives us an instance of the underlying type, which is numerical.
+#if DEBUG
+                    Type runtimeType = _generationSpec.MetadataLoadContext.Resolve(value.GetType());
+                    Debug.Assert(_generationSpec.IsNumberType(runtimeType));
+#endif
+
+                    // Return the numeric value.
+                    return $"({typeRef})({value.ToString()})";
+                }
+
+                switch (value)
+                {
+                    case string @string:
+                        return QuoteString(@string);
+                    case char @char:
+                        return String.Format("'{0}'", EscapeChar((char)value, inString: false));
+                    case double.NegativeInfinity:
+                        return "double.NegativeInfinity";
+                    case double.PositiveInfinity:
+                        return "double.PositiveInfinity";
+                    case double.NaN:
+                        return "double.NaN";
+                    case double @double:
+                        return $"({typeRef})({@double.ToString(JsonConstants.DoubleFormatString, CultureInfo.InvariantCulture)})";
+                    case float.NegativeInfinity:
+                        return "float.NegativeInfinity";
+                    case float.PositiveInfinity:
+                        return "float.PositiveInfinity";
+                    case float.NaN:
+                        return "float.NaN";
+                    case float @float:
+                        return $"({typeRef})({@float.ToString(JsonConstants.SingleFormatString, CultureInfo.InvariantCulture)})";
+                    case decimal.MaxValue:
+                        return "decimal.MaxValue";
+                    case decimal.MinValue:
+                        return "decimal.MinValue";
+                    case decimal @decimal:
+                        return @decimal.ToString(CultureInfo.InvariantCulture);
+                    case bool @bool:
+                        return ToCSharpKeyword(@bool);
+                    default:
+                        // Assume this is a number.
+                        return $"({typeRef})({value.ToString()})";
+                }
+            }
+
+            // The following logic to quote string constants with proper escaping is copied from the WriteMetadataConstant method from
+            // https://github.com/dotnet/arcade/blob/0cd94b1d02c03377d99f3739beb191591f6abee5/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Attributes.cs#L241
+
+            private static string QuoteString(string str)
```

Comment:
RegexGenerator uses:
https://github.com/dotnet/runtime/blob/55fecfbc201dc789d71c0903e04f2a779212c671/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs#L3342-L3343
Can this do the same?

Reply:
Awesome, thanks. JSON can rely on Roslyn here.

Reply:
FWIW I was hoping `SymbolDisplay.FormatPrimitive` could handle all cases needed by this source generator, but it doesn't handle some numeric types correctly. Opened https://github.com/dotnet/roslyn/issues/58705 to track a fix.